graphistry/ai_utils.py

```diff
@@ -127,3 +131,229 @@ def get_graphistry_from_milieu_search(
     ntdf = ndf[ndf[node_col].isin(gcols)]
     g = graphistry.edges(tdf, src, dst).nodes(ntdf, node_col)
     return g
+
+
+# #########################################################################################################################
+#
+#  Graphistry Vector Search Index
+#
+##########################################################################################################################
+
+
+def build_annoy_index(X, angular, n_trees=None):
+    """Builds an Annoy Index for fast vector search
+
+    Args:
+        X (_type_): _description_
+        angular (_type_): _description_
+        n_trees (_type_, optional): _description_. Defaults to None.
+
+    Returns:
+        _type_: _description_
+    """
+    from annoy import AnnoyIndex  # type: ignore
+
+    logger.info(f"Building Index of size {X.shape}")
+
+    if angular:
+        logger.info("-using angular metric")
+        metric = "angular"
+    else:
+        logger.info("-using euclidean metric")
+        metric = "euclidean"
+
+    search_index = AnnoyIndex(X.shape[1], metric)
+    # Add all the feature vectors to the search index
+    for i in range(len(X)):
+        search_index.add_item(i, X.values[i])
```

Comment:
Its sad that AnnoyIndex doesn't have something like .fit() to add everything at once.

Reply:
Agree -- but also amazing how fast this for loop works: would hope they had a vectorized version ...